daml-lf/spec/daml-lf-1.rst

```diff
@@ -763,11 +767,59 @@ as ``𝕋(F)``. See the `Built-in functions`_ section for the complete
 list of built-in functions and their respective types.
 
 
+Type synonym resolution
+.......................
+
+First, we define the synonym resolution relation ``↠`` over types,
+which inline type synonym definitions inside types::
+
+  ——————————————————————————————————————————————— RewriteVar
+   α  ↠  α
+
+  ——————————————————————————————————————————————— RewriteNat
+   n  ↠  n
+
+  ——————————————————————————————————————————————— RewriteBuiltin
+   BuiltinType ↠ BuiltinType
+
+  ———————————————————————————————————————————————— RewriteTyCon
+   Mod:T ↠  Mod:T
+
+   'synonym' S (α₁:k₁) … (αₙ:kₙ) ↦ τ  ∈ 〚Ξ〛Mod
+   τ  ↠  σ      τ₁  ↠  σ₁  ⋯  τₙ  ↠  σₙ
+  ——————————————————————————————————————————————— RewriteSynonym
+   Mod:S τ₁ … τₙ   ↠   σ[α₁ ↦ σ₁, …, αₙ ↦ σₙ]
+
+   τ₁ ↠ σ₁  ⋯  τₙ  ↠  σₙ
+  ———————————————————————————————————————————————— RewriteText
+   ⟨ f₁: τ₁, …, fₘ: τₘ ⟩ ↠ ⟨ f₁: σ₁, …, fₘ: σₘ ⟩
+
+   τ₁  ↠  σ₁        τ₂  ↠  σ₂
+  ———————————————————————————————————————————————— RewriteApp
+   τ₁ τ₂  ↠  σ₁ σ₂
```

Comment:
I think you need to split it out into two cases, in case `σ₁` happens to be a type synonym.

```
 τ₁  ↠  σ₁        τ₂  ↠  σ₂
σ₁ σ₂ is not a fully saturated type synonym application
———————————————————————————————————————————————— RewriteApp1
    τ₁ τ₂  ↠  σ₁ σ₂
```

 τ₁  ↠  σ₁        τ₂  ↠  σ₂
σ₁ σ₂ is a fully saturated type synonym application
σ₁ σ₂ ↠ σ 
———————————————————————————————————————————————— RewriteApp1
    τ₁ τ₂  ↠  σ
```

Reply:
Won’t we already all in the `RewriteSynonym` case for `RewriteApp2` here?

Reply:
Not if `σ₁` is a partially applied type synonym. Of course, that can't happen if type synonyms must be fully applied in every case.

Concrete example:

```
synonym Foo a b = a -> b 
synonym Bar = Foo Int64
```

then `Bar Int64 ` doesn't fall into either `RewriteSynonym` or `RewriteApp1`, you would need `RewriteApp2` to make progress with `Bar ↠ Foo Int64` before getting `Bar Int64 ↠ Int64 -> Int64`.